src/ROSInterface.cpp

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright (c) 2022, Daniel Beßler
+ * All rights reserved.
+ *
+ * This file is part of KnowRob, please consult
+ * https://github.com/knowrob/knowrob for license details.
+ */
+
+#include <termios.h>
+// STD
+#include <exception>
+#include <iostream>
+// BOOST
+#include <boost/program_options/options_description.hpp>
+#include <boost/program_options/variables_map.hpp>
+#include <boost/program_options/parsers.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+// KnowRob
+#include <knowrob/knowrob.h>
+#include <knowrob/Logger.h>
+#include <knowrob/HybridQA.h>
+// ROS
+#include <ros/ros.h>
+#include <ros/console.h>
+#include <ros/package.h>
+#include <json_prolog_msgs/PrologQuery.h>
+#include <json_prolog_msgs/PrologFinish.h>
+#include <json_prolog_msgs/PrologNextSolution.h>
+
+#define PARAM_INITIAL_PACKAGE "initial_package"
+#define PARAM_INITIAL_GOAL "initial_goal"
+#define PARAM_NUM_ROS_THREADS "num_ros_threads"
+#define PARAM_ENABLE_DEBUG "pl_debug"
+
+#define NUM_ROS_THREADS_DEFAULT 2
+
+using namespace knowrob;
+namespace po = boost::program_options;
+
+class HybridQAResultHandler : public QueryResultHandler {
+public:
+    bool incrementalMode;
+    std::string errorMessage;
+    std::list<std::shared_ptr<std::string>> solutions_;
+    HybridQAResultHandler(const boost::property_tree::ptree &config, const std::string &queryString, bool incremental = false)
+	: has_stop_request_(false),
+	  currentQuery_(""),
+      hybridQA_(config)
+	{
+        incrementalMode = incremental;
+        errorMessage = "";
+        currentQuery_ = queryString;
+        runQuery(currentQuery_);
```

Comment:
it seems strange that the "result handler" itself is issuing a query. it was supposed to only receive a solution. So the class name HybridQAResultHandler seems not appropriate.